src/plonk.rs

```diff
@@ -276,29 +287,48 @@ impl<C: HaloCurve> Circuit<C> {
             }
         }
 
-        // Reduce each opening set to a single point.
-        let opening_set_reductions: Vec<C::ScalarField> = all_opening_sets
-            .iter()
-            .map(|opening_set| {
-                C::ScalarField::inner_product(&actual_scalars, &opening_set.to_vec())
-            })
-            .collect();
-
-        // Then, we reduce the above opening set reductions to a single value.
-        let _reduced_opening = reduce_with_powers(&opening_set_reductions, v_sf);
-
         let u_scaling_bf = challenger.get_challenge();
         let u_scaling_sf = u_scaling_bf.try_convert::<C::ScalarField>()?;
         let u_curve = C::convert(u_scaling_sf) * self.u.to_projective();
-
         // Final IPA proof.
         let mut halo_a = reduced_coeffs;
-        let mut halo_b = powers(zeta_sf, self.degree());
+        // The Halo b vector is a random combination of the powers of all opening points.
+        let mut halo_b = self.build_halo_b(
+            &[
+                (0..2 * num_public_input_gates)
+                    .step_by(2)
+                    .map(|i| self.subgroup_generator_n.exp_usize(i))
```

Comment:
Probably not a significant cost, but this could also be `self.subgroup_n[i]` just to avoid the exp